### PR TITLE
Initial BEL and site design discussion.

### PR DESCRIPTION
**Pull request by litghost**
@vaughnbetz FYI, discussion of BEL boundary modelling of Stratix II and 10.

To view images inline, view here: https://github.com/litghost/fpga-interchange-schema/blob/add_bel_and_site_design/docs/bel_and_site_design.md

### CODE REVIEW

File: docs/bel_and_site_design.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+It is important to divide the place and route problem and the synthesis
+problem, at least as defined for the purpose of the FPGA interchange.  The
+synthesis tool operates on the **cell library**, which should be designed to
+expose logic elements at a useful level of granularity.
+
+As a concrete example, a LUT4 element is techinically just two LUT3 elements,
+connected by a mux (e.g. MUXF4), a LUT3 element is just two LUT2 elements,
+connected by a mux (e.g. MUXF3), etc. If the outputs of those interior muxes
+are not accessible to the place and route tool, then exposing those interior
+function muxes as cells in the cell library is not a useful.
```
Comment (gatecat) [inline docs/bel_and_site_design.md:23]: not as useful

File: docs/bel_and_site_design.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+Cell definitions should be granular enough that the synthesis can map to
+them, but not so granular that the place and route tool will be making few if
+any choices.  If there is only one legal placement of the cell, it's value is
+relatively low.
+
+## Drawing site boundries
+
+When designing an FPGA interchange device resource for a new fabric, one
```
Comment (gatecat) [inline docs/bel_and_site_design.md:32]: might want to add a specific note about FF control sets here (afaik a site boundary must contain all FFs sharing a control set)

**Comment by mithro** [comment]
Just FYI - the paper at https://www.eecg.utoronto.ca/~kmurray/titan/fpl_13_titan.pdf has some information about modeling the Stratix IV ALM.
Screenshot from 2021-04-08 08-22-37

**Comment by litghost** [comment]
> Just FYI - the paper at https://www.eecg.utoronto.ca/~kmurray/titan/fpl_13_titan.pdf has some information about modeling the Stratix IV ALM.

So the BEL design you've shown lumps the carry and the LUT element, which is not a good choice IMO for an interchange representation.  To wide of a boundry.

File: docs/bel_and_site_design.md (hunk 2 above)
Comment (litghost) [reply docs/bel_and_site_design.md:32]: Agreed, add a subsection talking both about the control set routing and non-routing type constraints.

File: docs/bel_and_site_design.md (hunk 1 above)
Comment (litghost) [reply docs/bel_and_site_design.md:23]: Fixed

File: docs/bel_and_site_design.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+drive them.  In this case, it is important to draw the site boundary large
+enough to capture these cases so as to enable the local congestion to be
+resolved during either packing for clustered approaches, or during placement
+during unclustered approaches.  In either case, local congestion that is
+strongly placement dependant must be resolved prior to general routing,
+unless a fused placement and routing algorithm is used.
+
+## Drawing BEL boundaries
+
+BEL definitions require that creating a boundary around primitive elements of
```
Comment (acomodi) [inline docs/bel_and_site_design.md:62]: Probably something is missing from this sentence

File: docs/bel_and_site_design.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+static routing muxes is the degree of configurability present in the
+underlying bitstream.  Some static routing muxes share configuration bits in
+the bitstream, and so expressing them as two seperate static routing muxes
+potentially gives the place and route tool flexibility than the underlying
+fabric cannot express.  This will result in physical netlists that cannot be
+converted to bitstream.
+
+In some cases this can be handled through tight coupling of the cell and
+BEL library.  The idea is to limit cell port to BEL pin mappings that avoid
+illegal static routing mux configurations.  This approach has it limits.
```
Comment (acomodi) [inline docs/bel_and_site_design.md:99]: its limits

File: docs/bel_and_site_design.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+So now consider the output paths from the the 4 LUT4 elements in the Stratix
+II site.  Some of the LUT4 outputs route directly to the carry element, so it
+will be important for the place and route tool be able to place a LUT4 or
+smaller to access that direct connection.  But if the output is not used in
+the carry element, then it can only be accessed in Stratix II via the MUXF5
+(blue below) and MUXF6 (red below) elements.
+
+![Stratix II Highlight MUXF5 and MUXF6](highlight_muxf5_muxf6.png)
+
+So given the Stratix II site layout, the following BELs will be requires:
```
Comment (acomodi) [inline docs/bel_and_site_design.md:138]: required

Comment (litghost) [reply docs/bel_and_site_design.md]: Fixed

File: docs/bel_and_site_design.md (hunk 4 above)
Comment (litghost) [reply docs/bel_and_site_design.md]: Fixed

File: docs/bel_and_site_design.md (hunk 3 above)
Comment (litghost) [reply docs/bel_and_site_design.md]: Fixed

File: docs/bel_and_site_design.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+ - 1 LUT6 BELs that connect to the output FF or output MUX
+
+### Versal ICAP
+
+The Versal ICAP LUT structure is fairly similiar to the Stratix 10 combitorial
+elements.
+
+![Versal ICAP LUTs](versal_luts.png)
+
+Unless the Stratix 10 ALM, it appears only 1 of the LUT4's connects to the
```
Comment (gatecat) [inline docs/bel_and_site_design.md:173]: unlike, not unless?

File: docs/bel_and_site_design.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+![Stratix 10 2 LUT5](stratix10_highlight_lut5.png)
+![Stratix 10 LUT6](stratix10_highlight_lut6.png)
+
+So given the Stratix 10 site layout, the following BELs will be required:
+
+ - 4 LUT4 BELs that connect to the carry
+ - 2 LUT5 BELs that connect to the output FF or output MUX
+ - 1 LUT6 BELs that connect to the output FF or output MUX
+
+### Versal ICAP
```
Comment (gatecat) [inline docs/bel_and_site_design.md:166]: ACAP, not ICAP

Comment (litghost) [reply docs/bel_and_site_design.md:166]: Opps, yes.

File: docs/bel_and_site_design.md (hunk 6 above)
Comment (litghost) [reply docs/bel_and_site_design.md]: Yes

**Review by gatecat — APPROVED** [review]

**Pull request merged**